### PR TITLE
[ONNX] Fix list dtype finding bug in dispatcher

### LINKED ISSUE #122166: [ONNX] Dispatcher does not handle list correctly when the first element is None
### 🐛 Describe the bug

When this dispatcher is handling inputs of type `Sequence[None | Tensor]`, it checks the type of only the first element. If the element is None it is unable to obtain the correct type:

https://github.com/pytorch/pytorch/blob/18d94d71659ee820e450c23ef1df3862396946ac/torch/onnx/_internal/fx/onnxfunction_dispatcher.py#L884

This leads to functions being selected incorrectly in the case of `aten::index`, where `[None, INT64, ...]` is dispatched to the `BOOL` overload.

cc @titaiwangms @BowenBao @fatcat-z 

### Versions

latest

### PR DESCRIPTION
**Pull request by titaiwangms**
Fixes #122166 

Before this PR, dispatcher assumes the first input should provide the reasonable dtype to them, but `aten::index` reveals the cases with `None` in the front of inputs. The PR addresses it by selecting the first non None input to take dtype.

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Have a test to validate this?

**Comment by titaiwangms** [comment]
> Have a test to validate this?

@justinchuby Do we have the updated `aten::index` included in 0315 onnxscript? If that's the case, we should have it in `test_fx_op_consistency.py`, and no need to add another add-hoc test.

**Comment by justinchuby** [comment]
We will need to check. The publishing pipeline is stalled.

File: torch/onnx/_internal/fx/onnxfunction_dispatcher.py [hunk 1]
```diff
@@ -881,7 +881,10 @@ def _find_onnx_data_type(
     if isinstance(torch_input, (int, float, bool, str, complex)):
         return fx_type_utils.from_torch_dtype_to_onnx_dtype_str(type(torch_input))
     if isinstance(torch_input, (list, tuple)) and torch_input:  # [Tensor, Tensor]
-        set_dtype = _find_onnx_data_type(torch_input[0])
+        the_first_non_none_item = next(
+            (item for item in torch_input if item is not None), None
+        )
+        set_dtype = _find_onnx_data_type(the_first_non_none_item)
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/onnxfunction_dispatcher.py:887]: What if it is all None? Should we match _any_ type in that case?

**Review by justinchuby — APPROVED** [review]
Thanks!

File: torch/onnx/_internal/fx/onnxfunction_dispatcher.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/onnxfunction_dispatcher.py:887]: It's going to hit the None if branch down below, which matches all types as the comments suggesting.

**Review by thiagocrepaldi — APPROVED** [review]
LGTM. But make sure **there is a test** to catch this behavior.

**Comment by justinchuby** [comment]
> > Have a test to validate this?
> 
> @justinchuby Do we have the updated `aten::index` included in 0315 onnxscript? If that's the case, we should have it in `test_fx_op_consistency.py`, and no need to add another add-hoc test.

Just checked. Yes the function is included in 3/15 onnxsciprt

**Pull request closed without merging**